lib/server_connection.ml

```diff
@@ -90,13 +90,22 @@ let on_wakeup_writer t k =
   then failwith "on_wakeup_writer on closed conn"
   else t.wakeup_writer := k::!(t.wakeup_writer)
 
-let _wakeup_writer callbacks =
-  let fs = !callbacks in
-  callbacks := [];
+let wakeup_writer t =
+  let fs = !(t.wakeup_writer) in
+  t.wakeup_writer := [];
   List.iter (fun f -> f ()) fs
 
-let wakeup_writer t =
-  _wakeup_writer t.wakeup_writer
+let rec _transfer_writer_callbacks fs reqd =
+  match fs with
+  | [] -> ()
+  | f :: fs ->
+    Reqd.on_more_output_available reqd f;
+    _transfer_writer_callbacks fs reqd
```

Comment:
Seems a little weird that this uses recursion and the above uses `List.iter`, but other than that it looks good

Reply:
This would allocate if it used `List.iter`, whereas the above doesn’t. The list itself should probably just be removed so only one callback can be registered at once. In which case this and the above will be simplified.

Reply:
Wow look at mr compiler over here.

I also think it shouldn't be a list. I keep having to trace through to understand where things get added to it, only to realize that it's the single continuation.